File: Algorithms/p257_Binary_Tree_Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        string cur_path;
        if(root) binaryTreePathsDFS(res,cur_path,root);
        return res;
    }
    void binaryTreePathsDFS(vector<string>& res, string& cur_path, TreeNode* root) {
        if(!cur_path.empty()) cur_path += "->";
        cur_path += to_string(root->val);
        if(!root->left&&!root->right){
            res.push_back(cur_path);
            return;
        }
        if(root->left){
            binaryTreePathsDFS(res,cur_path,root->left);
            cur_path.erase(cur_path.rfind('>')-1);
        }
        if(root->right){
            binaryTreePathsDFS(res,cur_path,root->right);
            cur_path.erase(cur_path.rfind('>')-1);
        }
    }
};

File: Algorithms/p257_Binary_Tree_Paths/p257_Binary_Tree_Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {string[]}
    def binaryTreePaths(self, root):
        res = []
        cur_path = []
        if root:
            self.binaryTreePathsDFS(res,cur_path,root)
        return res

    def binaryTreePathsDFS(self, res, cur_path, root):
        cur_path.append(str(root.val))
        if not root.left and not root.right:
            res.append('->'.join(cur_path))
            return
        if root.left:
            self.binaryTreePathsDFS(res,cur_path,root.left)
            cur_path.pop()
        if root.right:
            self.binaryTreePathsDFS(res,cur_path,root.right)
            cur_path.pop()
